Implemented enclave block store

byron-marohn opened the pull request:
The block store is a state cache for the eservice which eliminates the
need to send state back and forth on every transaction. State is pulled
from the block store when running contracts, and the resulting state is
placed there when the contract execution completes.

Current implementation details:
- The block store is not persistent across restarts
- The client must retrieve state from the block store to store on the
ledger after any operation that updates the state
- The client must put the state block(s) into the block store when
starting work with an enclave service that does not currently have them.

Signed-off-by: Byron Marohn <byron.marohn@intel.com>

byron-marohn commented:
Only one thing I'd like to fix before this gets merged - this causes warnings like this to show up in the test script: 
```swig/python detected a memory leak of type 'pdo_err_t *', no destructor found.```

The reason this happens is that the underlying implementation of BlockStoreGet (in the file `eservice/pdo/eservice/enclave/enclave/block_store.cpp`) allocates a buffer which is passed back to the caller (the enclave via ocall). It is clearly marked with a TODO comment. This is done because I don't know how else to pass back arbitrary size memory back in an ocall. Better ideas much appreciated! 

It wouldn't be too hard to make the swig error go away - since when calling from the SWIG side of things I can just have a different implementation which does not allocate memory (since ByteArray's are directly available when invoking via swig and not via ocall). It's only worth doing this work-around if we can't solve the problem for the ocall variant.

cmickeyb commented:
As far as the buffers go... there are several potential solutions... the most obvious (and least performant) is to just copy the buffer into a new location in untrusted space so that the buffer can be released on return. The second is to break it into two pieces... first ocall requests allocation of a buffer in untrusted space that is returned to the enclave, then either write directly into that buffer.

cmickeyb commented on eservice/pdo/eservice/enclave/block_store.h:
```diff
@@ -0,0 +1,36 @@
+void block_store_put(
+    const std::string& key,
+    const std::string& value
+    );
```
document the format expectations for all string parameters

cmickeyb commented on eservice/pdo/eservice/enclave/enclave/block_store.cpp:
```diff
@@ -0,0 +1,187 @@
+    // **********
+    // LOCK
+    pthread_spin_lock(&lock);
+
```
this is probably best moved into a module that manages the actual storage. That abstraction will simplify any move to persistent backup. and maybe the storage abstraction should be pushed into common?

byron-marohn replied:
I added a TODO for this - I think it's out of scope of the current commit. We should definitely replace the implementation of this - but I think adding yet another abstraction layer only to stuff this lame implementation in it is not worth doing.

cmickeyb commented on eservice/pdo/eservice/enclave/enclave/block_store.cpp:
```diff
@@ -0,0 +1,187 @@
+                            inValue.data(), inValue.size());
+
+    pdo::error::ThrowIf<pdo::error::ValueError>(
+       ret != 0, "Unable to put into the Block Store");
```
this is really the wrong kind of error to be throw at this point. Use ThrowPDOError

byron-marohn replied:
Fixed in upcoming commit.

cmickeyb commented on eservice/pdo/eservice/enclave/enclave/block_store.h:
```diff
@@ -0,0 +1,65 @@
+                size_t* valueSize
+                );
+
+            // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```
document parameter formats

byron-marohn commented on eservice/lib/libpdo_enclave/contract_response.cpp:
```diff
@@ -174,10 +176,23 @@ ByteArray ContractResponse::SerializeAndEncrypt(
+        if (sgx_ret != 0) {
+            SAFE_LOG(PDO_LOG_ERROR, "SGX error %d invoking ocall_BlockStorePut()", sgx_ret);
+            throw;
+        }
```
Fixed in upcoming commit. I have no idea where I copied this from but it's obviously wrong now that I look at it.

byron-marohn commented on eservice/lib/libpdo_enclave/contract_response.cpp:
```diff
@@ -174,10 +176,23 @@ ByteArray ContractResponse::SerializeAndEncrypt(
+        if (ret != 0) {
+            SAFE_LOG(PDO_LOG_ERROR, "Error %d saving state via ocall_BlockStorePut()", ret);
+            throw;
+        }
```
Fixed in upcoming commit.

byron-marohn commented on eservice/lib/libpdo_enclave/contract_state.cpp:
```diff
@@ -110,11 +112,36 @@ void ContractState::Unpack(const ByteArray& state_encryption_key_,
+            if (sgx_ret != 0) {
+                SAFE_LOG(PDO_LOG_ERROR, "SGX error %d invoking ocall_BlockStoreGet()", sgx_ret);
+                throw;
+            }
```
Fixed in upcoming commit.

byron-marohn commented on eservice/lib/libpdo_enclave/contract_state.cpp:
```diff
@@ -110,11 +112,36 @@ void ContractState::Unpack(const ByteArray& state_encryption_key_,
+            if (ret != 0) {
+                SAFE_LOG(PDO_LOG_ERROR, "Error %d retrieving state via ocall_BlockStoreGet()", ret);
+                throw;
+            }
```
Fixed in upcoming commit.

byron-marohn commented on eservice/lib/libpdo_enclave/block_store.edl:
```diff
@@ -0,0 +1,31 @@
+                                size_t keySize,
+                                [out] uint8_t** value,
+                                [out] size_t* valueSize);
+
```
Upcoming commit will replace this with a two-step operation - get size, then get data. This eliminates the memory leak / freeing problem and also reduces copying (uses the built-in SGX argument marshalling)

byron-marohn commented:
New commit addresses all the above comments - mostly two things: 
- Adding documentation to all the block store APIs
- Changing the ocall_BlockStoreGet to be a two-step operation - first an ocall_BlockStoreHead() to get the size, then the get. This enables use of the usual SGX argument marshalling code and removes an extra copy. It also fixes the (known about) memory leak. 

However, swig still thinks there is a problem someplace. I'm still seeing these messages in the log: 
```swig/python detected a memory leak of type 'pdo_err_t *', no destructor found.``` for tests that invoke the enclave directly, rather than going through the eservice. I have no idea what could be causing this.
